Friendbot as a Microservice

nikhilsaraf opened the pull request:
Extracted out Friendbot into it's own microservice so we can run it independently.

The friendbot microservice will call out to Horizon over HTTP/S, taking the horizon url from a new config file. Horizon will temporarily redirect all incoming requests (`/friendbot`) to the passed in friendbot_url at `/` along with the request query params.

This PR is a working separation of friendbot and is ready to merge from a functional perspective. It may be easier to review the code in the following logical order of the 3 commits.
- [x] Commit 1: Extract out common packages used by both Friendbot and Horizon out of `horizon/internal` (many changed files)
- [x] Commit 2: Extract Friendbot out of Horizon into it's own microservice
- [x] Commit 3: Redirect `/friendbot` on horizon to `/` on friendbot

------

In order to prevent merge conflicts from code drift and to keep the code more easily reviewable, I plan to have a second PR with additional improvements. This PR closes #134 and I will open a new issue for the leftover work outlined below.
- clean up middleware (logging, rate limiting, metrics)
- doc updates (horizon and friendbot)
- use [go-chi/chi](https://github.com/go-chi/chi) as mux

nikhilsaraf commented on services/friendbot/internal/friendbot.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// Pay funds the account at `destAddress`
+func (bot *Bot) Pay(destAddress string) (client.TransactionSuccess, error) {
```
reworked this file to use the client/horizon library to build a CreateAccount operation instead of using the txsub.

nikhilsaraf commented on services/friendbot/internal/friendbot_test.go:
```diff
@@ -1,27 +1,24 @@
-	"sync"
+	"github.com/stretchr/testify/assert"
 
-	"github.com/stellar/go/services/horizon/internal/test"
```
removed the horizon/internal/test dependency since we don't need it here

nikhilsaraf commented on services/friendbot/main.go:
```diff
@@ -0,0 +1,159 @@
+	PrivateKeyFile  string `toml:"private-key-file" valid:"required"`
+}
+
+// Config represents the configuration of a friendbot server
```
follows the command pattern in federation as opposed to what we do in horizon

nikhilsaraf commented on services/friendbot/main.go:
```diff
@@ -0,0 +1,159 @@
+
+// Web contains the http server related fields for horizon: the router,
+// rate limiter, etc.
+type Web struct {
```
this Web and initWeb will go away once we move to using the `go-chi/chi` mux. Used this for now since it was closely tied to our Action pattern and I wanted to keep that as a smooth extraction/refactor.

nikhilsaraf commented on services/friendbot/internal/actions_friendbot.go:
```diff
@@ -1,35 +1,36 @@
-		action.loadResource,
-
 		func() {
-			hal.Render(action.W, action.Resource)
```
since we use `hal`, I needed to move `horizon/internal/render` to `horizon/render`

nullstyle replied:
It's not acceptable for us to introduce cross-service dependencies in this way.  It's too easy IMO for service-specific decisions to become implicit dependencies between the two services to arise and for us to get painted into a corner or a painful refactoring.

Instead, if you'd like to leverage the hal library in friendbot, extract the non-horizon-specific parts to`support/render/hal`.  `PagingToken`,`Page`, `StandardPagingOptions`, and `PagedLink` shouldn't be extracted owing to the fact that they are too horizon-specific.  Also, it'd be great if you can update the tests to reflect our latest best practices.

nikhilsaraf commented on services/friendbot/internal/friendbot.go:
```diff
@@ -0,0 +1,103 @@
+	base64, err := txs.Base64()
+
+	// only increment the in-memory sequence number if we are going to submit the transaction, while we hold the lock
+	if err == nil {
```
small functional change here: I moved the bot sequence increment to be after we sign the txn as opposed to before we sign. this way we only increment if we will be sending to the network.

nikhilsaraf commented on services/horizon/internal/init_web.go:
```diff
@@ -50,15 +51,15 @@ func initWeb(app *App) {
 	r.Use(middleware.EnvInit)
 	r.Use(app.Middleware)
 	r.Use(middleware.RequestID)
-	r.Use(contextMiddleware(app.ctx))
```
extracted some middleware that is shared between horizon and friendbot. Happy to move this elsewhere if you want but we needed it for now so I moved it out of `horizon/internal`

nikhilsaraf commented on services/horizon/internal/init_web.go:
```diff
@@ -125,8 +126,12 @@ func initWebActions(app *App) {
 	// friendbot
-	r.Post("/friendbot", &FriendbotAction{})
-	r.Get("/friendbot", &FriendbotAction{})
+	redirectFriendbot := func(w http.ResponseWriter, r *http.Request) {
```
redirect operation from horizon

nikhilsaraf commented on services/friendbot/main.go:
```diff
@@ -0,0 +1,159 @@
+	r.Use(c.Handler)
+
+	// friendbot
+	r.Post("/", &internal.FriendbotAction{Friendbot: fb})
```
it was enabled in horizon. Horizon requests will be forwarded here so we still need to support POST requests.

nikhilsaraf commented on services/friendbot/main.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+func initRouter(r *web.Mux, fb *internal.Bot) {
+	ctx, _ := context.WithCancel(context.Background())
```
there's no context anymore

nikhilsaraf commented on services/friendbot/internal/friendbot_test.go:
```diff
@@ -1,27 +1,24 @@
 	_, err := fb.makeTx("GDJIN6W6PLTPKLLM57UW65ZH4BITUXUMYQHIMAZFYXF45PZVAWDBI77Z")
-
-	tt.Require.NoError(err)
+	assert.NoError(t, err)
```
done

nikhilsaraf commented:
addressed comments, there's only 1 item left in my books:
- [ ] friendbot error message should match current hal format to be consistent with existing API.

bartekn commented on services/friendbot/internal/friendbot.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	result, err := bot.Client.SubmitTransaction(signed)
+	if err != nil && strings.Contains(err.Error(), "tx_bad_seq") {
```
I think that `Error()` return value [does not](https://github.com/stellar/go/blob/master/clients/horizon/error.go#L11) contain the code. You should check `ResultCodes` method.

nikhilsaraf replied:
fixed

bartekn commented on services/friendbot/internal/friendbot.go:
```diff
@@ -0,0 +1,111 @@
+	return base64, err
+}
+
+// refreshes the sequence from the bot account and increments by 1
```
Outdated part of comment: `and increments by 1`

nikhilsaraf replied:
removed

bartekn commented on services/friendbot/internal/friendbot.go:
```diff
@@ -0,0 +1,111 @@
+	seq, err := strconv.ParseInt(botAccount.Sequence, 10, 0)
+	if err != nil {
+		return err
+	}
```
In case of errors in this function, would it help to reset sequence number to `0`? Then in `checkRefreshSequence` it will attempt to load it again instead of submitting tx with invalid one.

nikhilsaraf replied:
done

nikhilsaraf commented:
Good to go on my end, @nullstyle @bartekn

bartekn commented on services/friendbot/internal/friendbot.go:
```diff
@@ -0,0 +1,125 @@
+		case client.Error:
+			bot.checkHandleBadSequence(&e)
+		case *client.Error:
+			bot.checkHandleBadSequence(e)
```
Why? Shouldn't we always return the same error type?

nikhilsaraf replied:
the API returns type `error` although SubmitTransaction -> decodeResponse -> returns the pointer.
fixed.

bartekn commented on services/horizon/internal/init_web.go:
```diff
@@ -125,8 +126,12 @@ func initWebActions(app *App) {
+	redirectFriendbot := func(w http.ResponseWriter, r *http.Request) {
+		redirectURL := app.config.FriendbotURL + "?" + r.URL.RawQuery
+		http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
+	}
```
IMHO, we should remove friendbot completely and add path-based routing at proxy level.

nikhilsaraf replied:
yes, we should do that, was trying to reduce the ops overhead of making this dev change. This will still work and I can remove it once it's fully working via the proxy.

bartekn commented on support/http/server/factory.go:
```diff
@@ -0,0 +1,27 @@
+	router := make(map[string][]route)
+	for _, method := range supportedMethods {
+		router[method] = []route{}
+	}
```
We don't really need this. Maps hold default value (empty slice for slices) when there is no element yet. So `supportedMethods` is also unnecessary (we may want to add `DELETE` routes in a future).

nikhilsaraf replied:
ok, updated.

nikhilsaraf commented:
ping @nullstyle @bartekn

nikhilsaraf commented:
made all the suggested changes, @nullstyle 

for `halRender`, we will need to include that separation since the `Render` method moved there. This just adds 1 import and 1 rename for these files. Since in a few cases there are more than 1 calls to the original `hal` package, renaming that would cause more diffs. I'm happy to go with whatever route is easier for you to review.

nullstyle commented on services/friendbot/internal/friendbot_handler.go:
```diff
@@ -20,15 +19,15 @@ type FriendbotHandler struct {
 	result, err := handler.doHandle(r)
 	if err != nil {
-		problem.Render(log.DefaultLogger, w, err)
+		problem.Render(nil, w, err)
```
use `context.Background()` when no context is currently known.  However, you probably just want to use `r.Context()` in this situation.

nikhilsaraf commented:
done, @nullstyle

nullstyle commented on support/http/server/chi-router.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// Serve starts a web server by binding it to a socket and setting up the shutdown signals
+func Serve(router *chi.Mux, port int, tls *TLSConfig) {
```
I'm fine with us merging this code in the present state, but I think in general you should probably try to mirror the existing go http functions, keeping function signatures analogous when possible.  This method, for example, should probably be named `ListenAndServeTLS`, the router should be the last argument, and the port should probably just be a standard go listen string.

nullstyle commented on services/bifrost/server/server.go:
```diff
@@ -6,21 +6,20 @@ package server
 import (
```
I just want to point out that we already have server helpers in our `support/http` package:  https://github.com/stellar/go/blob/master/support/http/main.go#L36

nullstyle commented on support/http/server/middleware_logger.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// BindLoggerMiddleware returns a LoggerMiddleware bound to the passed in requestIdKey, thereby decoupling it from middleware.RequestID
+func BindLoggerMiddleware(requestIDKey interface{}) func(http.Handler) http.Handler {
```
Why do we need to bind the middleware to different request id keys?

nullstyle commented on support/http/server/middleware_logger.go:
```diff
@@ -0,0 +1,46 @@
+
+	next.ServeHTTP(ww, r)
+
+	duration := big.NewRat(
```
Doing big math to calculate duraction is overkill IMO.  Accuracy is not paramount... lets just use floating point math.

nullstyle replied:
Scratch the recommendation of floating math.  Let's copy how horizon's logger middleware does it:

https://github.com/stellar/go/blob/master/services/horizon/internal/middleware_logger.go#L32-L34

nullstyle commented on support/log/entry.go:
```diff
@@ -37,8 +38,10 @@ func (e *Entry) WithFields(fields F) *Entry {
 func (e *Entry) WithStack(stackProvider interface{}) *Entry {
```
I  We should probably expand this function to a type switch:

```go

func (e *Entry) WithStack(stackProvider interface{}) *Entry {
  var stack string
  switch sp := stackProvider.(type) {
    case errors.StackTracer:
      stack = fmt.Sprint(sp.StackTrace())
    case *gerr.Error:
      stack = fmt.Sprint(sp.ErrorStack())
    default:
      stack = "unknown"
  }

  return e.WithField("stack", stack)
}

```

nullstyle commented on support/log/main.go:
```diff
@@ -60,8 +56,11 @@ func Set(parent context.Context, logger *Entry) context.Context {
 // providing the default logger.
 func Ctx(ctx context.Context) *Entry {
-	found := ctx.Value(&contextKey)
+	if ctx == nil {
```
contexts should never be nil... that case represents a developer error, so IMO we should continue to panic if nil is provided.

nullstyle commented on support/http/server/middleware_logger.go:
```diff
@@ -0,0 +1,46 @@
+	"github.com/stellar/go/support/log"
+)
+
+func loggerMiddleware(requestIDKey interface{}, next http.Handler, w http.ResponseWriter, r *http.Request) {
```
My primary concern about this logging package is that there is no way to support horizon from it at present.  We include additional fields in the request/response log lines in horizon:

https://github.com/stellar/go/blob/master/services/horizon/internal/middleware_logger.go#L63

I'd like to have a story about how we are going to support horizon before we commit to a common package.

nullstyle commented:
I don't think any of the issues I've raised from this latest round are blockers, so I'm going to go ahead and approve a merge, but it'd be great if you could address the issues soon.
